rfcs/0001-inline-expression-value-logging.md

```diff
@@ -0,0 +1,191 @@
+
+## DSL extension of existing log macros
+
+Instead of adding a new set of _-v_ macros, it would be possible to
+extend the existing logging macros by using some additional marker
+syntax, such as the following:
+
+```rust
+debug!(= n/2)
+trace!("{} = {:x}", =i);
+```
+
+Here the `=` signals that the expression should be `stringify!`-ed for
+the message and its value returned from the macro. This complicates
+the macro's, but more importantly, considerably complicates the
+necessary guide documentation for new and existing users to understand
+an evolving logging _DSL_ as new syntax.  This syntax isn't any more
+compact. It aids comprehension when the macro arguments are as
+function-like as possible, by introducing the _-v_ macros specific to
+the feature as designed.
+
+## Only customize the value part of the format
+
+As proposed above, the entire format string may be customized. Until
+understanding that the single `i` argument is expanded to two
+arguments for formatting, it is surprising to see the following, as
+used above, with two placeholders in the format string:
+
+``` rust
+tracev!("{} = {:x}", i);
+```
+
+A workable alternative would be to only allow customizing the value
+part of the format, and have the expression part be fixed as `"{} =
+"`:
+
+``` rust
+tracev!("{:x}". i);
+```
+
+While this alternative is more compact for the subset of compatible
+use cases, it offers _less_ customization options.  For example,
+supplying a custom literal prefix to the message, for additional
+context. It is also less symmetric with the the _formatted logging_
+macros, where the format string literal represents the _entire_ text
+message of the log `Record`. Finally, this alternative adds a small
+but measurable amount of formatting overhead, as the value needs to be
+formatted to a temporary `fmt::Arguments` before formatting to the
+final message.
+
+## Release this as a separate crate, not in `log`
+
+The proposed additional logging macros play the same role as, for
+example, the existing `debug!` macro in `log`, which is just more
+convenient than using `log!(Level::Debug, …)` with an extra import.
+
+Ease of use was also an important part of the design and decision to
+add `dbg!` to rust `std` and the prelude.
+
+While the proposed additions would still require import for the
+macro(s), at least in 2018 edition projects, adding this feature to
+`log` avoids:
+
+* Needing to _discover_, add, and maintain an additional library
+  dependency. Discovery may be the biggest issue, and the inclusion of
+  `dbg!` in `std` and the prelude raises a high bar. This could be
+  partially mitigated by documented such an extension crate in the log
+  README and/or top-level library rustdoc.
+
+* The community effort to maintain such a separate library with
+  compatibility to the `log` crate, as it evolves.  For example, the
+  _-v_ macros will likely need to be adapted, when structured logging
+  is implemented ([log RFC 296]).
```

Comment:
How this interacts with structured logging will be interesting. I think we should be able to decorate a `Record` with at least the stringified expression as a structured value, and use the `Debug` representation of the value if the user hasn't supplied an explicit format function to `logv`.

Reply:
This is, in addition to allowing the same sort of structured key+values, at the end of the _-v_ macro arguments,?  Here *I think* you are providing evidence in favor of including this feature in _log_ itself, so these features can be more closely integrated as appropriate.

Reply:
Yeh, I was running through the exercise of whether we'd be worse off for adding this feature and then having to introduce structured logging without breakage to it later, which I think we'll be able to do just fine. So no issues there.

Reply:
Agreed. I touch on this a little in my below response to @dtolnay, going to resolve these here.